### PR TITLE
LUCENE-9810: Hunspell: when generating suggestions, skip too deep word FST subtrees

### PR DESCRIPTION
**Pull request by donnerpeter**
# Description

we skip roots longer than misspelled+4 anyway, so there's no need to read their arcs

# Solution

When we visit a node longer than needed, use `IntsRefFSTEnum#seekCeil` to speed up rewinding to the next shorter node.

# Tests

No new tests. `TestPerformance.de_suggest` has become ~40% faster

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java [hunk 1]
```diff
@@ -63,47 +65,57 @@
       String word, WordCase originalCase) {
     Comparator<Weighted<Root<String>>> natural = Comparator.naturalOrder();
     PriorityQueue<Weighted<Root<String>>> roots = new PriorityQueue<>(natural.reversed());
-    processFST(
```
Comment (donnerpeter) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/GeneratingSuggester.java:66]: processFST inlined

**Comment by donnerpeter** [comment]
This is quite a hot spot in suggestion generation. If you have other ideas how to speed up walking all dictionary entries, please share :)

**Comment by rmuir** [comment]
I'm not very familiar with the suggester here, but when I see optimizations like this, it reminds me of automata intersection with lucene's term dictionary (think wildcard/regex pattern queries, levenshtein fuzzy queries, and the like). Lucene's term dictionary has "next"/"advance" so it is basically an FSTEnum.

Maybe some tricks from that logic can give you ideas? https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/index/AutomatonTermsEnum.java

**Comment by donnerpeter** [comment]
Thanks! I'll think how to apply this. So far it seems that Hunspell's ngram algorithm is too automata-unfriendly :(

**Pull request merged**